edi_webservice_oca/tests/test_send_saving_response.py

```diff
@@ -0,0 +1,114 @@
+# Copyright 2023 Foodles (http://www.foodles.co).
+# @author Pierre Verkest <pierreverkest84@gmail.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+from contextlib import contextmanager
+from unittest import mock
+
+import responses
+
+from .common import TestEDIWebserviceBase
+
+
+class TestSend(TestEDIWebserviceBase):
+    @classmethod
+    def _setup_records(cls):
+        super()._setup_records()
+        cls.ws_backend = cls.backend.webservice_backend_id
+        cls.ws_backend.save_response = True
+        cls.settings = """
+        components:
+          send:
+            usage: webservice.send
+            work_ctx:
+              webservice:
+                method: post
+                kwargs:
+                  url_params:
+                    endpoint: push/here
+        """
+        cls.record.edi_exchange_state = "output_pending"
+
+        @contextmanager
+        def _consumer_record_no_new_env(self, record, new_cursor=True):
+            yield record
+
+        cls._consumer_record_no_new_env = _consumer_record_no_new_env
+
+    @responses.activate
+    def test_component_send_save_results_on_success(self):
+        self.record.type_id.set_settings(self.settings)
+        url = "https://foo.test/push/here"
+        response = "{'result': 'ok'}"
+        responses.add(responses.POST, url, body=response)
+        with mock.patch(
+            "odoo.addons.webservice.models.webservice_backend."
+            "WebserviceBackend._consumer_record_env",
+            side_effect=self._consumer_record_no_new_env,
+        ):
+            self.backend._check_output_exchange_sync()
+        self.assertEqual(self.record.edi_exchange_state, "output_sent")
+        self.assertEqual(
+            responses.calls[0].request.headers["Content-Type"], "application/xml"
+        )
+        self.assertEqual(responses.calls[0].request.body, "This is a simple file")
+        self.assertEqual(
+            self.record._get_file_content(field_name="ws_response_content"), response
+        )
+
+    @responses.activate
+    def test_component_send_save_results_on_http_error(self):
+        self.record.type_id.set_settings(self.settings)
+        url = "https://foo.test/push/here"
+        response = (
+            "{'error': 'something that details what went wrong to help investigations'}"
+        )
+        responses.add(
+            responses.POST,
+            url,
+            body=response,
+            status=401,
+        )
+
+        with mock.patch(
+            "odoo.addons.webservice.models.webservice_backend."
+            "WebserviceBackend._consumer_record_env",
+            side_effect=self._consumer_record_no_new_env,
+        ):
+            self.backend._check_output_exchange_sync()
+        self.assertEqual(self.record.edi_exchange_state, "output_error_on_send")
```

Comment:
Hum... since I've changed the behavior in `webserivce` to save data on the record in a different transaction, the state is not updated properly when HTTP error is in the _swallable_exceptions (Here the _consumer_record_env mock hide this issue as in the test there is only one transaction).

I've realized (thanks @simahawk ), this module can't take this decision, we have to let developers to choose at edi backend or exchange type level what's the expected behavior.

For instance in case of HTTP ERROR 500 we probably wants to retry the call and keep the edi exchange to  `output_pending`, in case of a web service returns some errors regarding the paylaods we probably want to move to `output_error_on_send`

So I'm going to stop `_swallable_exceptions` here and change `webservice` module in order to let developer change the state of the consumer exchange record.